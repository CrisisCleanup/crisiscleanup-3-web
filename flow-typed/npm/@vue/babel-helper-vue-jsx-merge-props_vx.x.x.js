// flow-typed signature: d7a23c53f4091ef2bf1df8afa5eb43c4
// flow-typed version: <<STUB>>/@vue/babel-helper-vue-jsx-merge-props_v^1.0.0/flow_v0.131.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@vue/babel-helper-vue-jsx-merge-props'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@vue/babel-helper-vue-jsx-merge-props' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@vue/babel-helper-vue-jsx-merge-props/dist/helper' {
  declare module.exports: any;
}

// Filename aliases
declare module '@vue/babel-helper-vue-jsx-merge-props/dist/helper.js' {
  declare module.exports: $Exports<'@vue/babel-helper-vue-jsx-merge-props/dist/helper'>;
}
