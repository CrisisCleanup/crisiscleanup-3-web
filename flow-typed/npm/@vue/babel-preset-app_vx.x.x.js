// flow-typed signature: 6ba91a92845d9a556a5bcdcc4f033fed
// flow-typed version: <<STUB>>/@vue/babel-preset-app_v^4.3.1/flow_v0.129.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@vue/babel-preset-app'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@vue/babel-preset-app' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@vue/babel-preset-app/polyfillsPlugin' {
  declare module.exports: any;
}

// Filename aliases
declare module '@vue/babel-preset-app/index' {
  declare module.exports: $Exports<'@vue/babel-preset-app'>;
}
declare module '@vue/babel-preset-app/index.js' {
  declare module.exports: $Exports<'@vue/babel-preset-app'>;
}
declare module '@vue/babel-preset-app/polyfillsPlugin.js' {
  declare module.exports: $Exports<'@vue/babel-preset-app/polyfillsPlugin'>;
}
