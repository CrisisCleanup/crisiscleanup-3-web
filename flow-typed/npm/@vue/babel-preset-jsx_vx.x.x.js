// flow-typed signature: eeecaeff29e4a3a6c63c4fe7e16a0eb3
// flow-typed version: <<STUB>>/@vue/babel-preset-jsx_v^1.1.2/flow_v0.131.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@vue/babel-preset-jsx'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@vue/babel-preset-jsx' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@vue/babel-preset-jsx/dist/plugin.cjs' {
  declare module.exports: any;
}

// Filename aliases
declare module '@vue/babel-preset-jsx/dist/plugin.cjs.js' {
  declare module.exports: $Exports<'@vue/babel-preset-jsx/dist/plugin.cjs'>;
}
