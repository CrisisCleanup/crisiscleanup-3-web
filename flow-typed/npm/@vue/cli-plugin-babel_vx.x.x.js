// flow-typed signature: 4cf1a98e80b05c0a167e013d64bc9645
// flow-typed version: <<STUB>>/@vue/cli-plugin-babel_v^4.4.6/flow_v0.131.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@vue/cli-plugin-babel'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@vue/cli-plugin-babel' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/customConfig.input' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/customConfig.output' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/default.input' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/default.output' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/doubleQuote.input' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/doubleQuote.output' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/require.input' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/require.output' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/templateLiteral.input' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/templateLiteral.output' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/codemods/usePluginPreset' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/generator' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/migrator' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-babel/preset' {
  declare module.exports: any;
}

// Filename aliases
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/customConfig.input.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/customConfig.input'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/customConfig.output.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/customConfig.output'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/default.input.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/default.input'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/default.output.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/default.output'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/doubleQuote.input.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/doubleQuote.input'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/doubleQuote.output.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/doubleQuote.output'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/require.input.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/require.input'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/require.output.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/require.output'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/templateLiteral.input.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/templateLiteral.input'>;
}
declare module '@vue/cli-plugin-babel/codemods/__testfixtures__/templateLiteral.output.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/__testfixtures__/templateLiteral.output'>;
}
declare module '@vue/cli-plugin-babel/codemods/usePluginPreset.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/codemods/usePluginPreset'>;
}
declare module '@vue/cli-plugin-babel/generator.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/generator'>;
}
declare module '@vue/cli-plugin-babel/index' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel'>;
}
declare module '@vue/cli-plugin-babel/index.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel'>;
}
declare module '@vue/cli-plugin-babel/migrator/index' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/migrator'>;
}
declare module '@vue/cli-plugin-babel/migrator/index.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/migrator'>;
}
declare module '@vue/cli-plugin-babel/preset.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-babel/preset'>;
}
