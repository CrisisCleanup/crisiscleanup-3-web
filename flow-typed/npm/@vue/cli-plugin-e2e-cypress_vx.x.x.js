// flow-typed signature: 122252e4d027555936ded476aff636b4
// flow-typed version: <<STUB>>/@vue/cli-plugin-e2e-cypress_v^4.4.6/flow_v0.129.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@vue/cli-plugin-e2e-cypress'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@vue/cli-plugin-e2e-cypress' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@vue/cli-plugin-e2e-cypress/generator' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/_eslintrc' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/plugins' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/specs/test' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/support/commands' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/support' {
  declare module.exports: any;
}

declare module '@vue/cli-plugin-e2e-cypress/ui' {
  declare module.exports: any;
}

// Filename aliases
declare module '@vue/cli-plugin-e2e-cypress/generator/index' {
  declare module.exports: $Exports<'@vue/cli-plugin-e2e-cypress/generator'>;
}
declare module '@vue/cli-plugin-e2e-cypress/generator/index.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-e2e-cypress/generator'>;
}
declare module '@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/_eslintrc.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/_eslintrc'>;
}
declare module '@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/plugins/index' {
  declare module.exports: $Exports<'@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/plugins'>;
}
declare module '@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/plugins/index.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/plugins'>;
}
declare module '@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/specs/test.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/specs/test'>;
}
declare module '@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/support/commands.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/support/commands'>;
}
declare module '@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/support/index' {
  declare module.exports: $Exports<'@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/support'>;
}
declare module '@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/support/index.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-e2e-cypress/generator/template/tests/e2e/support'>;
}
declare module '@vue/cli-plugin-e2e-cypress/index' {
  declare module.exports: $Exports<'@vue/cli-plugin-e2e-cypress'>;
}
declare module '@vue/cli-plugin-e2e-cypress/index.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-e2e-cypress'>;
}
declare module '@vue/cli-plugin-e2e-cypress/ui.js' {
  declare module.exports: $Exports<'@vue/cli-plugin-e2e-cypress/ui'>;
}
