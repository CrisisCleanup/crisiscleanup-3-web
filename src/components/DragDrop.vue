<template>
  <div class="flex items-center justify-center">
    <form ref="fileform" class="w-full h-full">
      <div
        class="relative w-full h-full flex flex-col items-center justify-center"
      >
        <slot>
          <div class="p-2 text-crisiscleanup-grey-700">{{ dragTitle }}</div>
          <div>{{ $t('or') }}</div>
          <div class="p-2 underline text-primary-dark">{{ chooseTitle }}</div>
        </slot>
        <input
          ref="fileinput"
          class="file-input"
          type="file"
          :multiple="multiple"
          :disabled="disabled"
        />
      </div>
    </form>
  </div>
</template>

<script>
export default {
  name: 'DragDrop',
  props: {
    chooseTitle: {
      type: String,
      default: '',
    },
    dragTitle: {
      type: String,
      default: '',
    },
    multiple: {
      type: Boolean,
      default: true,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      dragAndDropCapable: false,
    };
  },
  mounted() {
    this.dragAndDropCapable = this.determineDragAndDropCapable();
    if (this.dragAndDropCapable) {
      [
        'drag',
        'dragstart',
        'dragend',
        'dragover',
        'dragenter',
        'dragleave',
        'drop',
      ].forEach(evt => {
        this.$refs.fileform.addEventListener(
          evt,
          e => {
            e.preventDefault();
            e.stopPropagation();
          },
          false,
        );
      });

      const addDragOverClass = () => {
        this.$refs.fileform.classList.add('is-dragover');
      };

      const removeDragOverClass = () => {
        this.$refs.fileform.classList.remove('is-dragover');
      };

      this.$refs.fileform.addEventListener('drop', e => {
        for (let i = 0; i < e.dataTransfer.files.length; i++) {
          this.$emit('files', e.dataTransfer.files);
        }
        removeDragOverClass();
      });

      this.$refs.fileinput.addEventListener('change', e => {
        for (let i = 0; i < e.target.files.length; i++) {
          this.$emit('files', e.target.files);
        }
      });

      this.$refs.fileform.addEventListener('dragover', addDragOverClass, false);
      this.$refs.fileform.addEventListener(
        'dragenter',
        addDragOverClass,
        false,
      );
      this.$refs.fileform.addEventListener(
        'dragend',
        removeDragOverClass,
        false,
      );
      this.$refs.fileform.addEventListener(
        'dragleave',
        removeDragOverClass,
        false,
      );
    }
  },
  methods: {
    determineDragAndDropCapable() {
      const div = document.createElement('div');
      return (
        ('draggable' in div || ('ondragstart' in div && 'ondrop' in div)) &&
        'FormData' in window
      );
    },
  },
};
</script>

<style>
.file-drop-area {
  position: relative;
  display: flex;
  align-items: center;
  width: 500px;
  max-width: 100%;
  padding: 25px;
  border: 1px dashed black;
  border-radius: 3px;
  transition: 0.2s;
}

.fake-btn {
  flex-shrink: 0;
  background-color: rgba(255, 255, 255, 0.04);
  border: 1px solid rgba(255, 255, 255, 0.1);
  border-radius: 3px;
  padding: 8px 15px;
  margin-right: 10px;
  font-size: 12px;
  text-transform: uppercase;
}

.file-msg {
  font-size: small;
  font-weight: 300;
  line-height: 1.4;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.file-input {
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  width: 100%;
  cursor: pointer;
  opacity: 0;
}

.file-input:focus {
  outline: none;
}

.is-dragover {
  background-color: theme('colors.crisiscleanup-grey.200');
}
</style>
