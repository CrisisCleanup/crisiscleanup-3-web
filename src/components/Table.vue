<template>
  <div class="table-grid w-full">
    <div
      v-if="!hideHeader && $mq !== 'sm'"
      class="header text-crisiscleanup-grey-700 bg-white"
      :style="gridStyleHeader"
    >
      <div
        v-if="enableSelection && $mq !== 'sm'"
        class="flex items-center p-2 border-b"
      >
        <base-checkbox class="mb-5" @input="setAllChecked" />
      </div>
      <div
        v-if="hasRowDetails && $mq !== 'sm'"
        class="flex items-center p-2 border-b"
      ></div>
      <div
        v-for="column of columns"
        :key="column.key"
        class="p-2 border-b flex items-center cursor-pointer header-column"
        :class="column.headerClass || []"
        @click="
          () => {
            if (column.sortable) {
              sort(column.sortKey || column.key);
            }
          }
        "
      >
        <slot :name="`${column.key}-title`" :column="column">
          <base-text
            class="text-crisiscleanup-grey-700"
            :class="column.titleClass && column.titleClass"
            variant="h3"
            regular
          >
            {{ $t(column.title) }}
          </base-text>
        </slot>
        <div v-if="column.sortable">
          <ccu-icon
            v-if="
              sorter.key === (column.sortKey || column.key) &&
              sorter.direction === 'asc'
            "
            :alt="$t('actions.sort_ascending')"
            size="small"
            type="up"
          />
          <ccu-icon
            v-else-if="
              sorter.key === (column.sortKey || column.key) &&
              sorter.direction === 'desc'
            "
            :alt="$t('actions.sort_descending')"
            size="small"
            type="down"
          />
          <ccu-icon
            v-else
            :alt="$t('actions.sortable')"
            size="small"
            type="updown"
          />
        </div>
      </div>
      <template v-if="enableColumnSearch">
        <div
          v-for="column of columns"
          :key="`search-${column.key}`"
          class="
            p-2
            border-b
            flex
            items-center
            cursor-pointer
            bg-crisiscleanup-light-grey
          "
        >
          <template v-if="column.searchable">
            <form-select
              v-if="column.searchSelect"
              class="w-64 bg-white border h-10 border-crisiscleanup-dark-100"
              :options="column ? column.getSelectValues(data) : []"
              item-key="value"
              label="name_t"
              :value="columnSearch[column.key]"
              @input="
                (value) => {
                  columnSearch[column.key] = value;
                  onSearch();
                }
              "
            ></form-select>
            <base-input
              v-else
              :placeholder="column.title"
              :value="columnSearch[column.key]"
              @input="
                (value) => {
                  columnSearch[column.key] = value;
                  onSearch();
                }
              "
              input-style="width: 100%"
            ></base-input>
          </template>
        </div>
      </template>
    </div>
    <div class="body bg-white relative" :style="gridStyleBody">
      <div
        v-if="loading"
        class="
          absolute
          bottom-0
          left-0
          right-0
          top-0
          bg-crisiscleanup-light-grey
          opacity-75
          flex
          items-center
          justify-center
        "
      >
        <spinner />
      </div>
      <div
        v-for="item of data"
        :key="item.id"
        :style="gridStyleRow"
        class="hover:bg-crisiscleanup-light-grey"
        :class="{ 'bg-crisiscleanup-light-grey': selectedItems.has(item.id) }"
        @click="rowClick(item, $event)"
      >
        <div
          v-if="enableSelection && $mq !== 'sm'"
          class="flex items-center p-2 lg:border-b md:border-b"
        >
          <base-checkbox
            :value="selectedItems.has(item.id)"
            class="mb-5"
            data-cy="tableview_actionSelect"
            @input="
              (value) => {
                setChecked(item, value);
              }
            "
          />
        </div>
        <div
          v-if="hasRowDetails && $mq !== 'sm'"
          class="flex items-center p-2 lg:border-b md:border-b"
        >
          <font-awesome-icon
            class="cursor-pointer"
            size="md"
            :icon="showingDetails.has(item.id) ? 'caret-up' : 'caret-down'"
            @click="setShowingDetails(item)"
          />
        </div>
        <div
          v-for="column of columns"
          :key="column.key"
          class="flex items-center p-2 lg:border-b md:border-b cursor-pointer"
          :class="column.class || []"
          @click="handleColumnAction(column, item[column.key], item)"
        >
          <slot :name="column.key" :item="item">
            <span v-if="$mq === 'sm'" class="font-semibold mr-2">
              {{ column.title }}:
            </span>
            <template
              v-if="
                item[column.key] || item[column.key] === 0 || column.transformer
              "
            >
              <span v-if="column.transformer">{{
                column.transformer(item[column.key], item)
              }}</span>
              <span v-else>{{
                column.subKey
                  ? item[column.key][column.subKey]
                  : item[column.key]
              }}</span>
            </template>
          </slot>
        </div>
        <div
          v-if="hasRowDetails"
          v-show="showingDetails.has(item.id)"
          :style="`grid-column-start: 1; grid-column-end: ${
            columns.length + 1
          };`"
        >
          <slot name="rowDetails" :item="item"></slot>
        </div>
      </div>
      <div v-if="!data.length" class="p-2 text-crisiscleanup-grey-700 italic">
        {{ $t('info.no_items_found') }}
      </div>
    </div>
    <div
      v-if="enablePagination"
      class="footer flex items-center justify-between p-4"
    >
      <div class="flex items-center">
        <span class="mr-2">{{ $t('tableVue.per_page') }}</span>
        <form-select
          :value="pagination.pageSize"
          :options="pageSizes"
          :clearable="false"
          select-classes="w-24 bg-white border vue-select-up"
          @input="onSelectPageSize"
        />
      </div>
      <div>
        <div class="flex items-center">
          <base-button
            :disabled="isPreviousButtonDisabled"
            :action="
              () => {
                pageChangeHandle('previous');
              }
            "
            :text="$t('tableVue.prev')"
            :alt="$t('tableVue.prev')"
            icon="caret-left"
            class="mr-3 text-base"
          />
          <div>
            <template v-for="trigger in paginationTriggers">
              <span
                :key="trigger"
                :class="{
                  'rounded-full border px-3 py-1 bg-white shadow-inner':
                    trigger === pagination.current,
                }"
                class="cursor-pointer mx-4 text-base"
                @click="
                  () => {
                    pageChangeHandle(trigger);
                  }
                "
                >{{ trigger }}</span
              >
            </template>
          </div>
          <base-button
            :disabled="isNextButtonDisabled"
            :action="
              () => {
                pageChangeHandle('next');
              }
            "
            :text="$t('tableVue.next')"
            :alt="$t('tableVue.next')"
            suffix-icon="caret-right"
            class="text-base"
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { EventsMixin } from '@/mixins';
import { exportCSVFile } from '../utils/downloads';

export default {
  name: 'Table',
  mixins: [EventsMixin],
  props: {
    columns: {
      type: Array,
      default: () => {
        return [];
      },
    },
    hideHeader: {
      type: Boolean,
    },
    data: {
      type: null,
      default: null,
    },
    pagination: {
      type: Object,
      default: () => {
        return {};
      },
    },
    sorter: {
      type: Object,
      default: () => {
        return {};
      },
    },
    columnSearch: {
      type: Object,
      default: () => {
        return {};
      },
    },
    loading: Boolean,
    enableSelection: Boolean,
    enablePagination: Boolean,
    enableColumnSearch: Boolean,
    hasRowDetails: Boolean,
    bodyStyle: {
      type: Object,
      default: () => {
        return {};
      },
    },
  },
  data() {
    return {
      pageSizes: [5, 10, 20, 100, 500, 1000],
      visiblePagesCount: 5,
      selectedItems: new Set([]),
      showingDetails: new Set([]),
    };
  },
  computed: {
    pageCount() {
      return Math.ceil((this.pagination.total || 1) / this.pagination.pageSize);
    },
    isPreviousButtonDisabled() {
      return this.pagination.current === 1;
    },
    isNextButtonDisabled() {
      return this.pagination.current === this.pageCount;
    },
    paginationTriggers() {
      const currentPage = this.pagination.current;
      const { pageCount } = this;
      const { visiblePagesCount } = this;
      const visiblePagesThreshold = (visiblePagesCount - 1) / 2;
      const pagintationTriggersArray = Array(this.visiblePagesCount - 1).fill(
        0,
      );

      if (pageCount < visiblePagesCount) {
        return Array(pageCount)
          .fill(0)
          .map((_, index) => index + 1);
      }

      if (currentPage <= visiblePagesThreshold + 1) {
        pagintationTriggersArray[0] = 1;
        const pagintationTriggers = pagintationTriggersArray.map(
          (paginationTrigger, index) => {
            return pagintationTriggersArray[0] + index;
          },
        );
        pagintationTriggers.push(pageCount);
        return pagintationTriggers;
      }

      if (currentPage >= pageCount - visiblePagesThreshold + 1) {
        const pagintationTriggers = pagintationTriggersArray.map(
          (paginationTrigger, index) => {
            return pageCount - index;
          },
        );
        pagintationTriggers.reverse().unshift(1);
        return pagintationTriggers;
      }

      pagintationTriggersArray[0] = currentPage - visiblePagesThreshold + 1;
      const pagintationTriggers = pagintationTriggersArray.map(
        (paginationTrigger, index) => {
          return pagintationTriggersArray[0] + index;
        },
      );
      pagintationTriggers.unshift(1);
      pagintationTriggers[pagintationTriggers.length - 1] = pageCount;
      return pagintationTriggers;
    },
    gridTemplate() {
      const columnWidths = this.columns.map((column) => column.width || '1fr');
      if (this.hasRowDetails) {
        columnWidths.unshift('35px');
      }
      if (this.enableSelection) {
        columnWidths.unshift('35px');
      }
      return columnWidths.join(' ');
    },
    gridStyleHeader() {
      return {
        display: 'grid',
        'grid-template-columns': this.gridTemplate,
      };
    },
    gridStyleBody() {
      return {
        'max-height': '600px',
        overflow: 'auto',
        ...this.bodyStyle,
      };
    },
    gridStyleRow() {
      return {
        display: 'grid',
        'grid-template-columns': this.$mq === 'sm' ? 'auto' : this.gridTemplate,
      };
    },
  },
  mounted() {},
  methods: {
    setChecked(item, value) {
      if (value) {
        this.selectedItems.add(item.id);
      } else {
        this.selectedItems.delete(item.id);
      }
      this.selectedItems = new Set(this.selectedItems);
      this.$emit('selectionChanged', this.selectedItems);
    },
    setShowingDetails(item) {
      if (this.showingDetails.has(item.id)) {
        this.showingDetails.delete(item.id);
      } else {
        this.showingDetails.add(item.id);
      }
      this.showingDetails = new Set(this.showingDetails);
    },
    setAllChecked(value) {
      if (value) {
        this.selectedItems = new Set(this.data.map((item) => item.id));
        this.$emit('allSelected', this.selectedItems);
      } else {
        this.selectedItems = new Set();
        this.$emit('allDeselected', this.selectedItems);
      }
      this.$emit('selectionChanged', this.selectedItems);
    },
    pageChangeHandle(value) {
      let newPage;
      switch (value) {
        case 'next':
          newPage = this.pagination.current + 1;
          break;
        case 'previous':
          newPage = this.pagination.current - 1;
          break;
        default:
          newPage = value;
      }
      const columnSearch = { ...this.columnSearch };

      const pagination = {
        ...this.pagination,
        current: newPage,
        page: newPage,
      };
      const filter = {};
      const sorter = {
        ...this.sorter,
      };
      this.$emit('change', { pagination, filter, sorter, columnSearch });
    },
    onSelectPageSize(pageSize) {
      const columnSearch = { ...this.columnSearch };
      const pagination = {
        ...this.pagination,
        current: 1,
        pageSize,
      };
      const filter = {};
      const sorter = {
        ...this.sorter,
      };
      this.$emit('change', { pagination, filter, sorter, columnSearch });
    },
    sort(key) {
      const columnSearch = { ...this.columnSearch };
      const sorter = { ...this.sorter };
      sorter.key = key;
      if (sorter.direction === 'asc') {
        sorter.direction = 'desc';
      } else {
        sorter.direction = 'asc';
      }

      const pagination = {
        ...this.pagination,
      };
      const filter = {};
      this.logEvent(
        sorter.direction === 'asc' ? 'user_ui-sort-asc' : 'user_ui-sort-desc',
      );
      this.$emit('change', { pagination, filter, sorter, columnSearch });
    },
    onSearch() {
      const columnSearch = { ...this.columnSearch };
      const sorter = { ...this.sorter };
      const pagination = {
        ...this.pagination,
      };
      const filter = {};
      this.$emit('change', { pagination, filter, sorter, columnSearch });
    },
    rowClick(item, event) {
      const hasClass = (className) =>
        event.target.className.includes &&
        event.target.className.includes(className);
      if (
        event &&
        [
          'case-flag',
          'checkmark',
          'checkmark-input',
          'base-button',
          'link',
        ].some(hasClass)
      ) {
        return;
      }
      this.$emit('rowClick', item);
    },
    exportTableCSV() {
      const headers = {};
      const headersColumns = this.columns.filter((column) =>
        Boolean(column.title),
      );
      headersColumns.forEach((column) => {
        headers[column.key] = column.title;
      });

      const data = this.data.map((obj) => {
        const response = {};
        headersColumns.forEach((column) => {
          response[column.key] = obj[column.key];
        });
        return response;
      });
      exportCSVFile(headers, data, 'export');
    },
    async handleColumnAction(column, data, item) {
      if (column.action) {
        await column.action(data, item);
      }
    },
  },
};
</script>

<style scoped></style>
