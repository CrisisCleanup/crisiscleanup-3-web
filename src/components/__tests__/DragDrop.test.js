/**
 *
 * Tests for DragDrop
 *
 * Components
 */

import { shallowMount } from '@vue/test-utils';
import DragDrop from '@/components/DragDrop';
import Table from '../Table';

const mocks = {
  $t: (key) => key,
};

const mountWithOptions = () =>
  shallowMount(DragDrop, {
    stubs: {
      table: Table,
      'form-select': "<input type='select'>",
      'base-button': '<button></button>',
    },
    mocks,
  });

describe('DragDrop', () => {
  it('should not log any errors', () => {
    const spy = jest.spyOn(global.console, 'error');
    mountWithOptions();
    expect(spy).not.toHaveBeenCalled();
  });

  it('should render correctly and match snapshot', () => {
    const wrapper = mountWithOptions();
    expect(wrapper.element).toMatchSnapshot();
  });
});
