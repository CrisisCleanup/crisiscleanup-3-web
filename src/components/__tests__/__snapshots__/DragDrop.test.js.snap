// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragDrop should render correctly and match snapshot 1`] = `
<div
  class="flex items-center justify-center"
>
  <form
    class="w-full h-full"
  >
    <div
      class="relative w-full h-full flex flex-col items-center justify-center"
    >
      <div
        class="p-2 text-crisiscleanup-grey-700"
      >
        
      </div>
       
      <div>
        or
      </div>
       
      <div
        class="p-2 underline text-primary-dark"
      >
        
      </div>
       
      <input
        class="file-input"
        multiple="multiple"
        type="file"
      />
    </div>
  </form>
</div>
`;
